### ISSUE TITLE
Dynamic field parsing problem

### ISSUE THREAD

**Issue by ahives** [phatboyg/Machete#4]
I am calling this the dynamic field parsing problem because of the following usage scenario:

In the HL7 specification it states that the OBX-2 field's value determines how you will need to handle the OBX-5 field. This has been difficult in the past to deal with especially in statically generated specifications. In such cases, implementations usually represent both fields as string. However, traditionally the developer is left with the responsibility of having to parse out the OBX-5 field on their own. This means that parser is rather useless when it comes to OBX-5 if the developer is responsible for then building a mini parser on top. Consider the following example:

OBX|...|XPN|...|...|Doe^Jane^^^Dr^MD|...

So, in the above scenario the OBX-2 field is telling us that OBX-5 should be parsed as an XPN data type. In this case I would want to apply the parse to OBX-5 as a component field on extraction of the data rather than when the field is defined since you would not know in all cases what the data type is beforehand.

Perhaps the verb could something like:
`SelectAs(string dataTypeId)`

This would hang off of the field itself.

possible example,
`.Select<OBX>().Select(x => x.ObservationValue).SelectAs("XPN")`

This would, of course, blow up if no data type with the given ID is found.

**Comment by ahives** [comment]
We would probably want to do it with both a type and string representative of a type.

.As<XPN>()
.As("XPN")

The reason is that OBX-2 will have a strong representation and there way be use cases where we would know before hand and want to fail hard if it is not a particular type.

**Comment by phatboyg** [comment]
The schema will be needed, so it may be more along the lines of:

```
.Select<OBX>().Select(5, x => x.As<XPN>());
.Select<OBX>().Select(x => x.ObservationValue, x => x.As<XPN>());
```

**Comment by phatboyg** [comment]
The schema will be needed, so it may be more along the lines of:

```
.Select<OBX>().Select(5, x => x.As<XPN>());
.Select<OBX>().Select(x => x.ObservationValue, x => x.As<XPN>());
```

**Comment by ahives** [comment]
Would look something like this...

XPN val = result.Select(x => x.OBX).Select(x => x.ObservationValue).As("XPN");

**Comment by phatboyg** [comment]
Not really, the generic overloads would be messy. What I posted above can avoid the repeated generic type specifications.

**Comment by ahives** [comment]
You still need the ability to pass in a string representation of the type though.

**Comment by phatboyg** [comment]
Why? I don't understand the need for a non-strongly-typed argument.

**Comment by ahives** [comment]
There is a UI use case where I might want to define this through a drop down. What I am thinking is about a further down the line with having a DSL that will need to generate the proper code to do this. That is a use case that will be needed in the not too distant future. Any thoughts on how might this be done as to not shot ourselves in foot for future use cases like this.

**Comment by phatboyg** [comment]
So, thinking...

```
Value(x => x.ObservationValue, x =>
{
  x.SelectById(p => p.ValueType);
  x.Option<XPN>();
  x.Option<CWE>();
});
```

And then who knows.
